### PR TITLE
Noam merge

### PR DESCRIPTION
**Pull request by noamsabb**
-Game fully functionnal (few minor bugs left to fix)
-Re-Design all fxml file to re-design each window
-Added BoardScore and End-Game windows
-Added checkword func that will get us the actual word played and not only the tiles, taking care to add to the word the tile that was already on the screen
-Words can now be set on board in any order player want
-Errors handeling, pop-up window open to player in case an error happen letting him know and then letting him fix it

### CODE REVIEW

File: src/server/java/server/Game.java [hunk 1]
```diff
@@ -139,20 +140,25 @@ public void addPlayer(Player player) {
...
         sendGameToPlayer(player);
         sendPlayersToPlayers(player);
     }
 
-    public void playTurn(Player player, Word word) {
+    private void sendRounds(Player player) {
+        player.sendRounds(rounds);
+    }
+
+    public void playTurn(Player player, Word word, Word wordToReplaceW) {
```
Comment (edenfar) [inline src/server/java/server/Game.java:152]: You only need `Tiles[] tilesToReplace`

File: src/server/java/server/PlayerHandler.java [hunk 2]
```diff
@@ -91,13 +142,25 @@ private static Word parseMove(String move, Player player) {
...
             i++;
         }
-        return new Word(tiles, row, col, vertical);
+
+        Word word = new Word(tiles, row, col, vertical);
+        System.out.println(move);
+        return word;
+    }
+
+    public static String[] extractSubstrings(String input) {
```
Comment (edenfar) [inline src/server/java/server/PlayerHandler.java:161]: redundant function

**Review by edenfar — CHANGES_REQUESTED** [review]

File: src/client/java/org/bookscrabble/BoardController.java [hunk 3]
```diff
@@ -64,18 +89,20 @@ public BoardController(ViewModel vm) {
 
     }
 
+    public BoardController() {
```
[the remark this answers is not in the document]
Comment (noamsabb) [reply src/client/java/org/bookscrabble/BoardController.java:92]: We need a no arg controller use it in scoreBoard function line 393

File: src/server/java/server/PlayerHandler.java [hunk 4]
```diff
@@ -31,39 +61,61 @@ public void handleClient(InputStream inFromClient, OutputStream outToClient) {
...
-        while (!game.isOver()) {
-            request = in.nextLine();
-            if (request.equals("save"))
+        while (true) {
+            if (game.isOver()) {
+                game.endGame();
+                break;
+            }
+
+            String move = in.nextLine();
```
[the remark this answers is not in the document]
Comment (noamsabb) [reply src/server/java/server/PlayerHandler.java:97]: at this point its necessarily move, and we already have a request variable

File: src/client/java/viewmodel/ViewModel.java [hunk 5]
```diff
@@ -86,17 +101,24 @@ public void connectToGame() {
         this.isHost = false;
         GuestModel guestModel = new GuestModel();
         this.setModel(guestModel);
-        guestModel.connect("localhost", 6123);
-        guestModel.join(playerName.getValue(), gameName.getValue());
+
+        GamesManager gamesManager = GamesManager.get();
+        if (gamesManager.getGame(gameName.getValue()) != null) {
```
Comment (edenfar) [inline src/client/java/viewmodel/ViewModel.java:106]: This check should happen in the server

File: src/client/java/viewmodel/ViewModel.java [hunk 6]
```diff
@@ -86,17 +101,24 @@ public void connectToGame() {
...
     }
 
     public void onNewGame(String gameName) {
         this.gameName.setValue(gameName);
-        System.out.println(this.gameName.getValue());
     }
 
-    public void playTurn(String word, int r, int c, boolean vertical) {
-        m.playTurn(word, r, c, vertical);
+    public void playTurn(String word, int r, int c, boolean vertical, String wordToReplace) {
```
Comment (edenfar) [inline src/client/java/viewmodel/ViewModel.java:120]: `tilesToReplace`

File: src/server/java/server/Board.java
[the remark this answers is not in the document]
Comment (edenfar) [reply src/server/java/server/Board.java:1]: ?

File: src/server/java/server/PlayerHandler.java [hunk 7]
```diff
@@ -6,13 +6,15 @@
...
 
+
 public class PlayerHandler implements ClientHandler {
     /*
     All the communication with the player should happen here, and not in Player class
     We assume that the player has a thread listening to messages from the server (i.e., this class)
      */
     List<PrintWriter> outList = new ArrayList<>();
     List<Scanner> inList = new ArrayList<>();
+    Board board = new Board();
```
Comment (edenfar) [inline src/server/java/server/PlayerHandler.java:17]: You have the board of the game

**Review by edenfar — CHANGES_REQUESTED** [review]

File: src/server/java/server/Board.java [hunk 8]
```diff
@@ -272,18 +278,6 @@ public int tryPlaceWord(Word w) {
         return sum;
     }
 
-    public void print() {
```
Comment (edenfar) [inline src/server/java/server/Board.java:275]: why?

File: src/server/java/server/Game.java [hunk 9]
```diff
@@ -217,4 +229,46 @@ private void advanceCurrentPlayer() {
...
+        String playersScores[] = new String[numOfPlayers];
+        for (Player player : players) {
+            if (player != null) {
+                playersScores[players.indexOf(player)] = String.valueOf(player.getScore());
+            }
+        }
+        return playersScores;
+    }
+
+    public Board getGameBoard() {
```
Comment (edenfar) [inline src/server/java/server/Game.java:270]: `getBoard`

File: src/server/java/server/PlayerHandler.java [hunk 10]
```diff
@@ -91,28 +116,34 @@ private static Word parseMove(String move, Player player) {
 
         Tile[] tiles = new Tile[wordString.length()];
         int i = 0;
+
         for (char c : wordString.toCharArray()) {
-            if (player.getTile(c) == null)
-                throw new IllegalArgumentException("Player does not have tile " + c);
-            tiles[i] = player.getTile(c);
+            for (Tile t : p.getTiles()) {
```
Comment (edenfar) [inline src/server/java/server/PlayerHandler.java:121]: Use the function you wrote

File: src/server/java/server/PlayerHandler.java [hunk 11]
```diff
@@ -91,28 +116,34 @@ private static Word parseMove(String move, Player player) {
...
+                if (t.letter == c) {
+                    tiles[i] = p.getTile(c);
+                    break;
+                } else
+                    tiles[i] = g.getGameBoard().getTile(c);
+            }
             i++;
         }
-        return new Word(tiles, row, col, vertical);
+        Word word = new Word(tiles, row, col, vertical);
```
Comment (edenfar) [inline src/server/java/server/PlayerHandler.java:130]: just return it

**Review by edenfar — APPROVED** [review]

**Pull request merged**